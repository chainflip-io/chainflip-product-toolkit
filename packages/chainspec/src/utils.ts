import assert from 'assert';
import * as prettier from 'prettier';
import { type ParsedMetadata } from './BaseParser';

type ChangeType = 'added' | 'removed' | 'changed';
type Change = { path: string[]; type: ChangeType };

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(a: any, b: any, path: string[] = []): Generator<Change> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

class ChangelogGenerator {
  addedOrRemovedPallets: Record<string, 'added' | 'removed'> = {};

  changedEvents: Record<
    `${string}.${string}`,
    { field: string; type: 'added' | 'changed' | 'removed' }[]
  > = {};
  addedOrRemovedEvents: Record<`${string}.${string}`, 'added' | 'removed'> = {};

  trackedFieldChanges = new Set<string>();

  add({ path, type }: Change) {
    if (path.length === 1) {
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedPallets[path[0]] = type;
    } else if (path.length === 2) {
      const name = `${path[0]}.${path[1]}` as const;
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedEvents[name] = type;
    } else {
      const [pallet, event, , field] = path;
      const id = `${pallet}.${event}.${field}`;
      if (this.trackedFieldChanges.has(id)) return;
      this.trackedFieldChanges.add(id);
      (this.changedEvents[`${pallet}.${event}`] ??= []).push({ field, type });
    }
  }

  toString() {
    const lines: string[] = [];

    const pallets = Object.keys(this.addedOrRemovedPallets).sort();

    if (pallets.length) {
      lines.push('New or removed pallets:');
      for (const pallet of pallets) {
        lines.push(`  ${pallet}: ${this.addedOrRemovedPallets[pallet]}`);
      }
      lines.push('');
    }

    const events = Object.keys(this.addedOrRemovedEvents)
      .concat(Object.keys(this.changedEvents))
      .sort() as `${string}.${string}`[];

    let currentPallet = '';
    for (const event of events) {
      const [pallet, name] = event.split('.');
      if (pallet !== currentPallet) {
        if (currentPallet) lines.push('');
        currentPallet = pallet;
        lines.push(`${pallet}:`);
      }
      if (event in this.addedOrRemovedEvents) {
        lines.push(`  - ${name}: ${this.addedOrRemovedEvents[event]}`);
      } else {
        lines.push(`  - ${name}:`);
        const fields = this.changedEvents[event].map(
          ({ field, type }) => `    - ${field}: ${type}`,
        );
        lines.push(...fields);
      }
    }

    lines.push('');
    return lines.join('\n');
  }
}

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();
  const changelog = new ChangelogGenerator();

  for (const change of diff(a, b)) {
    changelog.add(change);

    const [pallet, event, field] = change.path;

    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return { changedOrAddedEvents, changelog: changelog.toString() };
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });
