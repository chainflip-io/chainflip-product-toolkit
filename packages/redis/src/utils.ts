type SnakeToCamelCase<T extends string> = T extends `${infer P}_${infer S}`
  ? `${P}${Capitalize<SnakeToCamelCase<S>>}`
  : T;

export const snakeToCamelCase = <const T extends string>(str: T): SnakeToCamelCase<T> =>
  str.replace(/_(.)/g, (match, group: string) => group.toUpperCase()) as SnakeToCamelCase<T>;

type CamelCasedKeys<T> =
  T extends Record<string, unknown>
    ? { [K in keyof T as K extends string ? SnakeToCamelCase<K> : K]: CamelCasedKeys<T[K]> }
    : T extends (infer U)[]
      ? CamelCasedKeys<U>[]
      : T;

export const transformKeysToCamelCase = <T>(obj: T): CamelCasedKeys<T> => {
  if (obj == null || typeof obj !== 'object') return obj as CamelCasedKeys<T>;

  if (Array.isArray(obj)) return obj.map(transformKeysToCamelCase) as CamelCasedKeys<T>;

  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => [
      snakeToCamelCase(key),
      transformKeysToCamelCase(value),
    ]),
  ) as CamelCasedKeys<T>;
};
