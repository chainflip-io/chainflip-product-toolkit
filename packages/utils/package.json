{
  "name": "@chainflip/utils",
  "version": "0.4.8",
  "type": "module",
  "scripts": {
    "eslint:check": "pnpm eslint --max-warnings 0 './**/*.ts'",
    "prettier:base": "prettier ./** --ignore-path=../../.prettierignore",
    "prettier:check": "pnpm prettier:base --check",
    "prettier:write": "pnpm prettier:base --write",
    "clean": "rm -rf dist",
    "prepublish": "pnpm test run && pnpm build:clean",
    "build": "tsup",
    "build:clean": "pnpm clean && pnpm build",
    "test:ci": "CI=1 pnpm t run",
    "test": "vitest",
    "coverage": "vitest run --coverage"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "exports": {
    "./*": {
      "require": {
        "types": "./dist/*.d.cts",
        "default": "./dist/*.cjs"
      },
      "import": {
        "types": "./dist/*.d.ts",
        "default": "./dist/*.js"
      }
    }
  },
  "devDependencies": {
    "@noble/hashes": "^1.5.0",
    "@vitest/ui": "2.1.4"
  },
  "dependencies": {
    "@date-fns/utc": "^2.1.0",
    "bignumber.js": "^9.1.2",
    "date-fns": "4.1.0"
  },
  "repository": "https://github.com/chainflip-io/chainflip-product-toolkit.git",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  }
}
