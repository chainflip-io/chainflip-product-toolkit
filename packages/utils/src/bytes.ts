import { assert } from './assertion';
import { type HexString } from './types';

export const bytesToHex = (input: Uint8Array | number[]): `0x${string}` => {
  const bytes = new Uint8Array(input);

  return `0x${Array.from(bytes)
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('')}`;
};

export const hexToBytes = (input: HexString): Uint8Array => {
  assert(/^0x[\da-f]*$/i.test(input) && input.length % 2 === 0, 'Invalid hex string');

  const hex = input.slice(2);

  const bytes = new Uint8Array(hex.length / 2);

  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = Number.parseInt(hex.slice(i, i + 2), 16);
  }

  return bytes;
};

const convertBase = (bytes: Uint8Array | number[], fromBase: number, toBase: number): number[] => {
  const result = [];

  for (const byte of bytes) {
    let carry = byte;

    for (let i = 0; i < result.length; i += 1) {
      carry += result[i] * fromBase;
      result[i] = carry % toBase;
      carry = Math.floor(carry / toBase);
    }

    while (carry !== 0) {
      result.push(carry % toBase);
      carry = Math.floor(carry / toBase);
    }
  }

  let leadingZeros = 0;
  while (bytes[leadingZeros] === 0) {
    leadingZeros += 1;
    result.push(0);
  }

  return result.reverse();
};

export const encodeBytesWithCharset = (bytes: Uint8Array | number[], charset: string): string =>
  convertBase(bytes, 256, charset.length)
    .map((charCode) => charset.charAt(charCode))
    .join('');

export const decodeBytesWithCharset = (input: string, charset: string): Uint8Array => {
  assert(new RegExp(`^[${charset}]*$`).test(input), 'Invalid input');

  const charMap = Object.fromEntries([...charset].map((char, index) => [char, index]));

  const bytes = input.split('').map((char) => charMap[char]);

  return new Uint8Array(convertBase(bytes, charset.length, 256));
};

const addPrefix = (input: string): HexString => {
  const [, bytes] = /^(?:0x)?([a-f\d]*)$/i.exec(input) || [];
  assert(bytes, 'Invalid hex string');
  return `0x${bytes}`;
};

export function reverseBytes(input: HexString): HexString;
export function reverseBytes(input: string): string;
export function reverseBytes(input: string): string {
  const reversed = bytesToHex(hexToBytes(addPrefix(input)).reverse());
  return input.startsWith('0x') ? reversed : reversed.slice(2);
}
