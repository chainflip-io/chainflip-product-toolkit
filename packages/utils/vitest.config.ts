import { coverageConfigDefaults } from 'vitest/config';
import config from '../../vitest.config';

export default {
  ...config,
  test: {
    ...config.test,
    coverage: {
      ...config.test?.coverage,
      exclude: ['**/consts.ts', '**/chainflip.ts', ...coverageConfigDefaults.exclude],
      thresholds: {
        lines: 90,
        statements: 90,
        branches: 90,
        functions: 90,
      },
    },
  },
};
