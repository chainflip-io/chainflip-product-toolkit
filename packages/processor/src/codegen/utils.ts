import { Queue } from '@chainflip/utils/async';
import * as fs from 'fs/promises';
import * as path from 'path';
import * as prettier from 'prettier';
import { z } from 'zod';
import { ParsedMetadata } from './Parser';
import assert from 'assert';

type ChangeType = 'added' | 'removed' | 'changed';
type Change = { path: string[]; type: ChangeType };

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(a: any, b: any, path: string[] = []): Generator<Change> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

class ChangelogGenerator {
  addedOrRemovedPallets: Record<string, 'added' | 'removed'> = {};

  changedEvents: Record<
    `${string}.${string}`,
    { field: string; type: 'added' | 'changed' | 'removed' }[]
  > = {};
  addedOrRemovedEvents: Record<`${string}.${string}`, 'added' | 'removed'> = {};

  trackedFieldChanges = new Set<string>();

  add({ path, type }: Change) {
    if (path.length === 1) {
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedPallets[path[0]] = type;
    } else if (path.length === 2) {
      const name = `${path[0]}.${path[1]}` as const;
      assert(type === 'added' || type === 'removed');
      this.addedOrRemovedEvents[name] = type;
    } else {
      const [pallet, event, , field] = path;
      const id = `${pallet}.${event}.${field}`;
      if (this.trackedFieldChanges.has(id)) return;
      this.trackedFieldChanges.add(id);
      (this.changedEvents[`${pallet}.${event}`] ??= []).push({ field, type });
    }
  }

  toString() {
    const lines: string[] = [];

    const pallets = Object.keys(this.addedOrRemovedPallets).sort();

    if (pallets.length) {
      lines.push('New or removed pallets:');
      for (const pallet of pallets) {
        lines.push(`  ${pallet}: ${this.addedOrRemovedPallets[pallet]}`);
      }
      lines.push('');
    }

    const events = Object.keys(this.addedOrRemovedEvents)
      .concat(Object.keys(this.changedEvents))
      .sort() as `${string}.${string}`[];

    let currentPallet = '';
    for (const event of events) {
      const [pallet, name] = event.split('.');
      if (pallet !== currentPallet) {
        if (currentPallet) lines.push('');
        currentPallet = pallet;
        lines.push(`${pallet}:`);
      }
      if (event in this.addedOrRemovedEvents) {
        lines.push(`  - ${name}: ${this.addedOrRemovedEvents[event]}`);
      } else {
        lines.push(`  - ${name}:`);
        const fields = this.changedEvents[event].map(
          ({ field, type }) => `    - ${field}: ${type}`,
        );
        lines.push(...fields);
      }
    }

    lines.push('');
    return lines.join('\n');
  }
}

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();
  const changelog = new ChangelogGenerator();

  for (const change of diff(a, b)) {
    changelog.add(change);

    const [pallet, event, field] = change.path;

    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return { changedOrAddedEvents, changelog: changelog.toString() };
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });

const cacheSchema = z.record(
  z.object({
    hash: z.string(),
    network: z.enum(['mainnet', 'perseverance', 'sisyphos', 'backspin', 'localnet']),
  }),
);
export type Network = z.output<typeof cacheSchema>[string]['network'];

export class SpecVersionCache {
  private contents?: z.output<typeof cacheSchema>;

  private readonly queue = new Queue();

  constructor(private readonly path: string) {}

  async read(): Promise<z.output<typeof cacheSchema>> {
    const contents = JSON.parse(
      await fs.readFile(this.path, 'utf8').catch(async () => {
        await fs.mkdir(path.dirname(this.path), { recursive: true });
        return '{}';
      }),
    ) as unknown;

    this.contents ??= cacheSchema.parse(contents);

    return this.contents;
  }

  async write(id: number, hash: string, network: Network): Promise<void> {
    return this.queue.enqueue(async () => {
      const data = await this.read();
      data[id] = { hash, network };
      await fs.writeFile(this.path, JSON.stringify(data, null, 2));
    });
  }

  async getVersion(hash: string, network: Network) {
    const data = await this.read();
    const [version] =
      Object.entries(data).find(([, d]) => d.hash === hash && d.network === network) ?? [];
    return version !== undefined ? Number(version) : undefined;
  }
}

export const specVersionCache = new SpecVersionCache(
  path.join(import.meta.dirname, '..', '..', 'metadata', 'specVersion.json'),
);

export const networkToRpcUrl: Record<Network, string> = {
  mainnet: 'https://rpc.mainnet.chainflip.io',
  perseverance: 'https://archive.perseverance.chainflip.io',
  sisyphos: 'https://archive.sisyphos.chainflip.io',
  backspin: 'https://backspin-rpc.staging',
  localnet: 'http://127.0.0.1:9944',
};
