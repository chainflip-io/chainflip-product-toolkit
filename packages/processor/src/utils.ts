import * as path from 'path';
import * as prettier from 'prettier';
import * as fs from 'fs/promises';
import { z } from 'zod';
import { ParsedMetadata } from './Parser';
import { Queue } from '@chainflip/utils/async';

/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
// a function to deep diff two metadata objects
function* diff(
  a: any,
  b: any,
  path: string[] = [],
): Generator<{ path: string[]; type: 'added' | 'removed' | 'changed' }> {
  const addedKeys = Object.keys(b)
    .filter((key) => b[key] !== undefined)
    .filter((key) => !a[key]);
  const removedKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => !b[key]);
  const commonKeys = Object.keys(a)
    .filter((key) => a[key] !== undefined)
    .filter((key) => b[key]);

  for (const key of addedKeys) {
    yield { path: [...path, key], type: 'added' };
  }

  for (const key of removedKeys) {
    yield { path: [...path, key], type: 'removed' };
  }

  for (const key of commonKeys) {
    if (typeof a[key] === 'object' && typeof b[key] === 'object') {
      yield* diff(a[key], b[key], [...path, key]);
    } else if (a[key] !== b[key]) {
      yield { path: [...path, key], type: 'changed' };
    }
  }
}
/* eslint-enable @typescript-eslint/no-unsafe-member-access */
/* eslint-enable @typescript-eslint/no-unsafe-argument */

export const diffSpecs = (a: ParsedMetadata, b: ParsedMetadata) => {
  const changedOrAddedEvents = new Set<string>();

  for (const change of diff(a, b)) {
    const [pallet, event, field] = change.path;
    if (event && (field || change.type === 'added')) {
      // if a field changed or a new event was added
      changedOrAddedEvents.add(`${pallet}.${event}`);
    } else if (change.type === 'added') {
      // if a new pallet was added
      Object.keys(b[pallet])
        .map((e) => `${pallet}.${e}`)
        .forEach((ev) => changedOrAddedEvents.add(ev), changedOrAddedEvents);
    }
  }

  return changedOrAddedEvents;
};

export const formatCode = (code: string) =>
  prettier.format(code, {
    parser: 'typescript',
    singleQuote: true,
    trailingComma: 'all',
    printWidth: 100,
  });

const cacheSchema = z.record(
  z.object({
    hash: z.string(),
    network: z.enum(['mainnet', 'perseverance', 'sisyphos', 'backspin']),
  }),
);
export type Network = z.output<typeof cacheSchema>[string]['network'];

export class SpecVersionCache {
  private contents?: z.output<typeof cacheSchema>;

  private readonly queue = new Queue();

  constructor(private readonly path: string) {}

  async read(): Promise<z.output<typeof cacheSchema>> {
    const contents = JSON.parse(
      await fs.readFile(this.path, 'utf8').catch(async () => {
        await fs.mkdir(path.dirname(this.path), { recursive: true });
        return '{}';
      }),
    ) as unknown;

    this.contents ??= cacheSchema.parse(contents);

    return this.contents;
  }

  async write(id: number, hash: string, network: Network): Promise<void> {
    return this.queue.enqueue(async () => {
      const data = await this.read();
      data[id] = { hash, network };
      await fs.writeFile(this.path, JSON.stringify(data, null, 2));
    });
  }

  async getVersion(hash: string, network: Network) {
    const data = await this.read();
    const [version] =
      Object.entries(data).find(([, d]) => d.hash === hash && d.network === network) ?? [];
    return version !== undefined ? Number(version) : undefined;
  }
}

export const specVersionCache = new SpecVersionCache(
  path.join(import.meta.dirname, '..', 'metadata', 'specVersion.json'),
);

export const networkToRpcUrl: Record<Network, string> = {
  mainnet: 'https://mainnet-archive.chainflip.io',
  perseverance: 'https://archive.perseverance.chainflip.io',
  sisyphos: 'https://archive.sisyphos.chainflip.io',
  backspin: 'https://backspin-rpc.staging',
};
