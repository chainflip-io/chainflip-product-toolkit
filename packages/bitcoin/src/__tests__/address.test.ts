import { bytesToHex } from '@chainflip/utils/bytes';
import { describe, it, expect } from 'vitest';
import { encodeAddress, decodeAddress, isValidAddressForNetwork } from '../index';

const networks = ['mainnet', 'perseverance', 'sisyphos', 'backspin', 'regtest', 'testnet'] as const;

const cases = (
  [
    ['P2WPKH', '0x63b7cc4432bba5c51fd09428c47abaf2d52f9373'],
    ['P2SH', '0x86bc2bfa0fdb414245561008c08615f4632604f9'],
    ['P2PKH', '0xecd30fe4f116f7485d7fd2f1cf26d0a059561548'],
    ['P2WSH', '0x321e6259700f0734a61b5ffa849de17865bc3293d5d069d91f23e6cb87daedf0'],
    ['Taproot', '0x68a3db628eea903d159131fcb4a1f6ed0be6980c4ff42b80d5229ea26a38439e'],
  ] as const
).flatMap(([type, address]) => networks.map((n) => [type, n, address] as const));

describe('bitcoin', () => {
  it.each(cases)('encodes and decodes a %s address on %s', (kind, network, address) => {
    const result = encodeAddress(address, kind, network);
    expect(result).toMatchSnapshot(`${kind} on ${network}`);
    expect(encodeAddress(Buffer.from(address.slice(2), 'hex'), kind, network)).toEqual(result);
    expect(encodeAddress([...Buffer.from(address.slice(2), 'hex')], kind, network)).toEqual(result);
    expect(bytesToHex(decodeAddress(result, network).data)).toEqual(address);
  });

  it('throws for invalid networks', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f9373', 'P2WPKH', 'invalid' as any),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: Invalid network: invalid]`);
  });

  it('throws for invalid address types', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f9373', 'invalid' as any, 'mainnet'),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: Invalid address type: invalid]`);
  });

  it('throws for invalid addresses', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f937', 'P2WPKH', 'mainnet'),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: bytes must have an even number of characters]`);
  });

  it('throws for non-hex addresses', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f937g', 'P2WPKH', 'mainnet'),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: bytes are not a valid hex string]`);
  });
});

describe(isValidAddressForNetwork, () => {
  const keys = <T extends Record<string, unknown>>(obj: T) => Object.keys(obj) as (keyof T)[];

  const bitcoinAddresses = {
    mainnet: {
      P2PKH: ['1PYVSoeftFP4EVBN3ou8vZctkhDthJamvp'],
      P2SH: ['32k55FA93MYqbjhLk9hokD3P666Vz9QqKb'],
      Taproot: ['bc1qwqdg6squsna38e46795at95yu9atm8azzmyvckulcc7kytlcckxswvvzej'],
      P2WSH: ['bc1qeklep85ntjz4605drds6aww9u0qr46qzrv5xswd35uhjuj8ahfcqgf6hak'],
    },
    testnet: {
      P2PKH: ['mipcBbFg9gMiCh81Kj8tqqdgoZub1ZJRfn', 'n2ZtW11kqP6pf3umiU4NaSRgrL9aQJVuJV'],
      P2SH: ['2MzQwSSnBHWHqSAqtTVQ6v47XtaisrJa1Vc'],
      Taproot: ['tb1qw508d6qejxtdg4y5r3zarvary0c5xw7kxpjzsx'],
    },
    regtest: {
      Taproot: ['bcrt1qs758ursh4q9z627kt3pp5yysm78ddny6txaqgw'],
    },
  };

  it('handles mainnet addresses', () => {
    const addresses = bitcoinAddresses.mainnet;

    for (const kind of keys(addresses)) {
      for (const address of addresses[kind]) {
        expect(isValidAddressForNetwork(address, 'mainnet')).toBe(true);
      }
    }

    for (const network of ['testnet', 'regtest'] as const) {
      for (const address of Object.values(bitcoinAddresses[network]).flat()) {
        expect(isValidAddressForNetwork(address, 'mainnet')).toBe(false);
      }
    }
  });

  it('handles testnet addresses', () => {
    const addresses = bitcoinAddresses.testnet;

    for (const kind of keys(addresses)) {
      for (const address of addresses[kind]) {
        expect(isValidAddressForNetwork(address, 'testnet')).toBe(true);
      }
    }

    for (const network of ['mainnet', 'regtest'] as const) {
      for (const address of Object.values(bitcoinAddresses[network]).flat()) {
        expect(isValidAddressForNetwork(address, 'testnet')).toBe(false);
      }
    }
  });

  it('handles regtest addresses', () => {
    const addresses = bitcoinAddresses.regtest;

    for (const kind of keys(addresses)) {
      for (const address of addresses[kind]) {
        expect(isValidAddressForNetwork(address, 'regtest')).toBe(true);
      }
    }

    for (const address of Object.values(bitcoinAddresses.mainnet).flat()) {
      expect(isValidAddressForNetwork(address, 'regtest')).toBe(false);
    }

    for (const type of keys(bitcoinAddresses.testnet)) {
      for (const address of bitcoinAddresses.testnet[type]) {
        expect(isValidAddressForNetwork(address, 'regtest')).toBe(type !== 'Taproot');
      }
    }
  });

  it('returns false for invalid addresses', () => {
    expect(isValidAddressForNetwork('invalid', 'mainnet')).toBe(false);
    expect(
      isValidAddressForNetwork(
        'bc1teklep85ntjz4605drds6aww9u0qr46qzrv5xswd35uhjuj8ahfcqgf6hak',
        'mainnet',
      ),
    ).toBe(false);
  });
});
