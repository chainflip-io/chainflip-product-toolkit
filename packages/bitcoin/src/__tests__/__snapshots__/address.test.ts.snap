// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bitcoin > encodes and decodes a P2PKH address on backspin > P2PKH on backspin 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > encodes and decodes a P2PKH address on mainnet > P2PKH on mainnet 1`] = `"1NbDK88PF7P5B4wTVk3HcJ3U8GJeiSrccq"`;

exports[`bitcoin > encodes and decodes a P2PKH address on perseverance > P2PKH on perseverance 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > encodes and decodes a P2PKH address on regtest > P2PKH on regtest 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > encodes and decodes a P2PKH address on sisyphos > P2PKH on sisyphos 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > encodes and decodes a P2PKH address on testnet > P2PKH on testnet 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > encodes and decodes a P2SH address on backspin > P2SH on backspin 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > encodes and decodes a P2SH address on mainnet > P2SH on mainnet 1`] = `"3DyRteshezqPfEwP5gyNX77nmyomu5GW2U"`;

exports[`bitcoin > encodes and decodes a P2SH address on perseverance > P2SH on perseverance 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > encodes and decodes a P2SH address on regtest > P2SH on regtest 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > encodes and decodes a P2SH address on sisyphos > P2SH on sisyphos 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > encodes and decodes a P2SH address on testnet > P2SH on testnet 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > encodes and decodes a P2WPKH address on backspin > P2WPKH on backspin 1`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > encodes and decodes a P2WPKH address on mainnet > P2WPKH on mainnet 1`] = `"bc1qvwmuc3pjhwju287sjs5vg7467t2jlymnmjyatp"`;

exports[`bitcoin > encodes and decodes a P2WPKH address on perseverance > P2WPKH on perseverance 1`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > encodes and decodes a P2WPKH address on regtest > P2WPKH on regtest 1`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > encodes and decodes a P2WPKH address on sisyphos > P2WPKH on sisyphos 1`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > encodes and decodes a P2WPKH address on testnet > P2WPKH on testnet 1`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > encodes and decodes a P2WSH address on backspin > P2WSH on backspin 1`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > encodes and decodes a P2WSH address on mainnet > P2WSH on mainnet 1`] = `"bc1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcqamg47k"`;

exports[`bitcoin > encodes and decodes a P2WSH address on perseverance > P2WSH on perseverance 1`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > encodes and decodes a P2WSH address on regtest > P2WSH on regtest 1`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > encodes and decodes a P2WSH address on sisyphos > P2WSH on sisyphos 1`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > encodes and decodes a P2WSH address on testnet > P2WSH on testnet 1`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > encodes and decodes a Taproot address on backspin > Taproot on backspin 1`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > encodes and decodes a Taproot address on mainnet > Taproot on mainnet 1`] = `"bc1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qxt7wp4"`;

exports[`bitcoin > encodes and decodes a Taproot address on perseverance > Taproot on perseverance 1`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > encodes and decodes a Taproot address on regtest > Taproot on regtest 1`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > encodes and decodes a Taproot address on sisyphos > Taproot on sisyphos 1`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > encodes and decodes a Taproot address on testnet > Taproot on testnet 1`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;
